src/kvtree_util.c

```diff
@@ -225,3 +228,119 @@ int kvtree_util_get_ptr(const kvtree* hash, const char* key, void** val)
 
   return rc;
 }
+
+/*
+ * Given a prefix passed to a kvtree_write_gather() (like "/tmp/rank2file",
+ * read all the rank data from all the scattered subfiles, and construct a
+ * kvtree in 'data'.  'data' will look something like:
+ *
+ *       26
+ *         FILE
+ *           ckpt.1/rank_26.ckpt
+ *       24
+ *         FILE
+ *           ckpt.1/rank_24.ckpt
+ *       25
+ *         FILE
+ *           ckpt.1/rank_25.ckpt
+ *       18
+ *         FILE
+ *           ckpt.1/rank_18.ckpt
+ *       ...
+ *
+ *  ... with each top-level element being the rank number.
+ */
+int kvtree_read_scatter_single(const char* prefix, kvtree* data)
+{
+  int rc = KVTREE_SUCCESS;
+  unsigned long level, expected_ranks, actual_ranks = 0;
+  char *prefix_dir, *prefix_dir_copy = NULL, *prefix_file, *prefix_file_copy = NULL;
+  char tmp[PATH_MAX];
+  kvtree *subfile_tree = NULL, *subfile_rank_tree = NULL;
+  kvtree *final_tree = kvtree_new();
+
+  /*
+   * Look at the high level kvtree and get the total number of ranks to
+   * expect.
+   */
+  rc = kvtree_read_file(prefix, final_tree);
+  if (rc != KVTREE_SUCCESS) {
+    return rc;
+  }
+  rc = kvtree_util_get_unsigned_long(final_tree, "RANKS", &expected_ranks);
+  if (rc != KVTREE_SUCCESS) {
+    kvtree_free(&final_tree);
+    return rc;
+  }
+  kvtree_delete(&final_tree);
+
+  /* Start construction of the final tree that we will return */
+  final_tree = kvtree_new();
+
+  /* Look at all the subfiles with our prefix and read in the file lists */
+  prefix_dir_copy = strdup(prefix);
+  prefix_dir = dirname(prefix_dir_copy);
+  prefix_file_copy = strdup(prefix);
+  prefix_file = basename(prefix_file_copy);
+
+  DIR *d;
+  struct dirent *dir;
+  d = opendir(prefix_dir);
+  if (!d) {
+    rc = KVTREE_FAILURE;
+    goto end;
+  }
+
+  /* For each file/dir in our prefix dir */
+  while ((dir = readdir(d)) != NULL) {
+    /*
+     * Search for any file starting with "prefix_file."
+     * like "rank2file.".
+     */
+    snprintf(tmp, sizeof(tmp), "%s.", prefix_file);
```

Comment:
Let's strengthen this check to use a regex on something like ``^prefix.0.\d+$``.  This helps avoid collisions on files that have similar prefix names, e.g., ``rank2file`` and ``rank2file.foo`` or even harder ``rank2file.0``